New rule: Meta refresh has no delay, no exception (#1527)
act-rules/act-rules.github.io: _rules/meta-refresh-no-delay-no-exception-bisz58.md

name: '`meta` element has no refresh delay (no exception)'
rule_type: atomic
description: |
  This rule checks that the `meta` element is not used for delayed redirecting or refreshing.
accessibility_requirements:
  wcag20:2.2.4: # Interruptions (AAA)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag20:3.2.5: # Change on Request (AAA)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:G110: # Using an instant client-side redirect
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:H76: # Using meta refresh to create an instant client-side redirect
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
htmlHintIgnore:
  # https://www.npmjs.com/package/htmlhint
  # (used with `npm test` to ensure validity of code snippets)
  - 'title-require'
---

## Applicability

> This rule applies to the first [valid](https://html.spec.whatwg.org/#attr-meta-http-equiv-refresh) `<meta http-equiv="refresh">` element with a `content` attribute in a document.

Reviewer: I think we should update that description of the `meta` element to a more modern writing style.

## Expectation

> The `time` of the `content` attribute is 0.

Reviewer: This too is a fairly dated way of writing expectations.

> See the definition of the [Refresh state (`http-equiv="refresh"`)](https://html.spec.whatwg.org/#attr-meta-http-equiv-refresh) for a precise description on how to determine the `time`.

Reviewer 2: This looks like a note without the note tag. We should move it to the background section.

## Assumptions

- This rule assumes that no functionality was provided by the website for the user to adjust the timer.

## Accessibility Support

> Not all major web browsers parse the value of the `content` attribute in the same way. This can cause redirects to happen in some browsers, but not in others. This can cause some pages to pass this rule, while still having a redirect in a minority of web browsers.

Reviewer: I think this should be more explicit about what they do differently. Otherwise we're just leaving people guessing.

Author: I honestly don't know 😅 
Since you are an author of the original rule, do you remember what this AS note is referring to?

Reviewer 3: I'll just leave this here: https://github.com/act-rules/act-rules.github.io/pull/1422/files. Not that I don't agree 🤷

Reviewer 2: I don't recall it either. But is it really needed to say anything more than "some browsers redirect while others don't"? I don't think we need to explain why some browsers don't follow the spec.

Reviewer: Mark has a couple examples: https://github.com/act-rules/act-rules.github.io/issues/1141#issue-558511928. From the look of it, If Firefox can't parse the value it default to a 0s delay. So:

```suggestion
Not all major web browsers parse the value of the `content` attribute in the same way. Some major browsers, if they are unable to parse the fault default to a 0s delay, whereas others will not redirect at all. This can cause some pages to be inapplicable for this rule, while still having a redirect in a minority of web browsers.
```

> ## Background

Reviewer 2: We should add the the links to the multiple understanding documents here.

- [G110: Using an instant client-side redirect](https://www.w3.org/WAI/WCAG21/Techniques/general/G110)
- [H76: Using meta refresh to create an instant client-side redirect](https://www.w3.org/TR/WCAG-TECHS/H76.html)
- [F40: Failure of Success Criterion 2.2.1 and 2.2.4 due to using meta redirect with a time limit](https://www.w3.org/TR/WCAG-TECHS/F40.html)
- [F41: Failure of Success Criterion 2.2.1, 2.2.4, and 3.2.5 due to using meta refresh with a time-out](https://www.w3.org/TR/WCAG-TECHS/F41.html)

## Test Cases

### Passed

#### Passed Example 1

This `meta` element redirects the user immediately. Users won't notice the change in context.

```html
<head>
	<meta http-equiv="refresh" content="0; URL='https://w3c.org'" />
</head>
```

#### Passed Example 2

The first valid `meta` element redirects immediately.

```html
<head>

> 	<meta http-equiv="refresh" content="0; http://w3c.org" />

Reviewer 2: ```suggestion
	<meta http-equiv="refresh" content="0; https://w3c.org" />
```

> 	<meta http-equiv="refresh" content="5; http://w3c.org" />

Reviewer 2: ```suggestion
	<meta http-equiv="refresh" content="5; https://w3c.org" />
```

</head>
```

### Failed

#### Failed Example 1

This `meta` element refreshes the page after 30 seconds.

```html
<head>
	<meta http-equiv="refresh" content="30" />
</head>
```

#### Failed Example 2

This `meta` element redirects the user after 30 seconds.

```html
<head>
	<meta http-equiv="refresh" content="30; URL='https://w3c.org'" />
</head>
```

#### Failed Example 3

The first `meta` element is not valid (because of the colon instead of a semi-colon in the `content` attribute), the second one redirects after 5 seconds.

```html
<head>

> 	<meta http-equiv="refresh" content="0: http://w3c.org" />

Reviewer 2: ```suggestion
	<meta http-equiv="refresh" content="0: https://w3c.org" />
```

> 	<meta http-equiv="refresh" content="5; http://w3c.org" />

Reviewer 2: ```suggestion
	<meta http-equiv="refresh" content="5; https://w3c.org" />
```

accessibility_requirements:
  wcag20:2.2.4: # Interruptions (AAA)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag20:3.2.5: # Change on Request (AAA)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:G110: # Using an instant client-side redirect
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:H76: # Using meta refresh to create an instant client-side redirect
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
acknowledgments:
  authors:
    - Jean-Yves Moyen
htmlHintIgnore:
  # https://www.npmjs.com/package/htmlhint
  # (used with `npm test` to ensure validity of code snippets)
  - 'title-require'
---

## Applicability

This rule applies to the first `meta` element in a document for which all the following are true:

- the element has an `http-equiv` [attribute value][] of `"refresh"`; and

> - the element has a [valid][valid content] `content` [attribute value][].

Author: I'm not super happy with the link "valid content" which goes to an **usage** of the `content` term rather than a proper anchor. It's unstable (if usages of `content` are added or removed). But there is no anchor on that specific part of the specs…
Ideas?

Reviewer 2: Having our own definition (that would copy the spec)?
That would take care of the instability issue, but we would need to update the definition if the spec gets updated.

## Expectation

> For each test target, running the [shared declarative refresh steps][], given the target's document, the value of the target's `content` attribute, and the target results in _time_ being either 0 or greater than 72000 (20 hours).

Author: Not a great fan of the call to the steps either. Didn't really find a better way to get the _time_ value… This seems like a huge stretch for "attribute value of time".
Ideas?

Reviewer 2: Not really :-(

> For each test target, running the [shared declarative refresh steps][], given the target's document, the value of the target's `content` attribute, and the target results in _time_ being either 0 or greater than 72000 (20 hours).

Reviewer 2: I don't think this rule should have the 20 hours exception, since this is the "no exception" rule (and the 20 hours exception comes from SC 2.2.1 which is not checked in this rule).

Author: 🙈 copy/pasting for the win (or loss)!
